src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/ExtensionMembersDataFlow.cs

```diff
@@ -0,0 +1,135 @@
+        static void TestExtensionPropertyAnnotatedReturn()
+        {
+            GetWithFields().ExtensionMembersPropertyAnnotatedReturn.RequiresPublicMethods();
+        }
+
+
+        [ExpectedWarning("IL2072", "GetWithMethods", "ExtensionMembersProperty")]
+        [ExpectedWarning("IL2072", "ExtensionMembersProperty", "RequiresPublicFields")]
+        static void TestExtensionPropertyMismatch()
+        {
+            GetWithMethods().ExtensionMembersProperty.RequiresPublicFields();
+        }
+
+        [ExpectedWarning("IL2026", "ExtensionMembersPropertyRequires")]
+        static void TestExtensionPropertyRequires()
+        {
+            _ = GetWithFields().ExtensionMembersPropertyRequires;
+        }
+
+        [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)]
+        static Type GetWithFields() => null;
+
+        [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
+        static Type GetWithMethods() => null;
+    }
+
+    [ExpectedNoWarnings]
+    public static class ExtensionMembers
+    {
+        extension([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] Type type)
+        {
+            public void ExtensionMembersMethod() => type.RequiresPublicFields();
+
+            [ExpectedWarning("IL2067", "RequiresPublicMethods")] // The attribute gets copied to the internal class too, but that one is unused. Then it complains.
+            public void ExtensionMembersMethodMismatch() => type.RequiresPublicMethods();
+
+            [RequiresUnreferencedCode(nameof(ExtensionMembersMethodRequires))]
+            public void ExtensionMembersMethodRequires() { }
+
+            [ExpectedWarning("IL2067", "type", "RequiresPublicConstructors")]
+            [ExpectedWarning("IL2067", "typeParam", "RequiresPublicMethods")]
+            public void ExtensionMembersMethodWithParams([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] Type typeParam)
+            {
+                type.RequiresPublicConstructors();
+                typeParam.RequiresPublicMethods();
+            }
+
+            [RequiresUnreferencedCode(nameof(ExtensionMembersStaticMethodRequires))]
+            public static void ExtensionMembersStaticMethodRequires() { }
+
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
+            public Type ExtensionMembersProperty => null;
+
+            public Type ExtensionMembersPropertyAnnotatedReturn
+            {
+                [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
+                get => null;
+            }
+
+            public Type ExtensionMembersPropertyRequires
+            {
+                [RequiresUnreferencedCode("ExtensionMembersPropertyRequires")]
+                get => null;
+            }
+        }
```

Comment:
What should we do if we find another annotation on the method, implying another requirement on the 'this' parameter? If we want to combine the requirement, it would require more demangling compiler generated code in the trimmer, so I don't think we should allow that, but we should have a test for it.

This also compiles to DAM on a static method, which doesn't make sense. Do we warn if we see DAM on a static method? If so we should eventually add a warning to the analyzer when it sees this.


```suggestion
            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
            public void ThisParameterIsAlsoAnnotated()
            {
                type.RequiresPublicFields(); // Definitely shouldn't warn.
                type.RequiresPublicMethods(); // Don't warn?
        }
```